ksql-streams/src/main/java/io/confluent/ksql/execution/streams/GroupByParamsFactory.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.codegen.ExpressionMetadata;
+import io.confluent.ksql.execution.util.StructKeyUtil;
+import io.confluent.ksql.execution.util.StructKeyUtil.KeyBuilder;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.util.SchemaUtil;
+import java.util.List;
+import java.util.function.Function;
+import org.apache.kafka.connect.data.Struct;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class GroupByParamsFactory {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GroupByParamsFactory.class);
+
+  private static final String GROUP_BY_VALUE_SEPARATOR = "|+|";
+
+  private GroupByParamsFactory() {
+  }
+
+  public static GroupByParams build(
+      final LogicalSchema sourceSchema,
+      final List<ExpressionMetadata> expressions
+  ) {
+    final Function<GenericRow, Struct> mapper = expressions.size() == 1
+        ? new SingleExpressionGrouper(expressions.get(0))::apply
+        : new MultiExpressionGrouper(expressions)::apply;
+
+    final LogicalSchema schema = expressions.size() == 1
+        ? singleExpressionSchema(sourceSchema, expressions.get(0).getExpressionType())
```

Comment:
Where do we validate that this type is a primitive?

Reply:
I'll be checking that all works nicely in https://github.com/confluentinc/ksql/issues/4113